relayer/src/chain/cosmos/wait.rs

```diff
@@ -50,26 +50,16 @@ pub async fn wait_tx_hashes(
     timeout: &Duration,
     tx_hashes: &[TxHash],
 ) -> Result<Vec<TxSearchResponse>, Error> {
-    let mut join_handles = Vec::new();
-
-    for tx_hash in tx_hashes.iter() {
-        let rpc_client = rpc_client.clone();
-        let rpc_address = rpc_address.clone();
-        let timeout = *timeout;
-        let tx_hash = *tx_hash;
-
-        let handle = task::spawn(async move {
-            wait_tx_hash(&rpc_client, &rpc_address, &timeout, &tx_hash).await
-        });
-        join_handles.push(handle);
-    }
-
-    let mut responses = Vec::new();
-
-    for handle in join_handles.into_iter() {
-        let response = handle.await.map_err(Error::join)??;
-        responses.push(response);
-    }
+    let tasks = tx_hashes
+        .iter()
+        .map(|tx_hash| async { wait_tx_hash(rpc_client, rpc_address, timeout, tx_hash).await })
+        .collect::<FuturesOrdered<_>>();
+
+    let responses = tasks
+        .collect::<Vec<_>>()
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?;
```

Comment:
A bit hard to follow the logic here. The first `.collect` makes sure all futures are executed, whereas `.try_collect` would bail out after one of them resolved an error. Then, the resolved vector of `Result` from all tasks is transformed to determine if the overall result is a success, anyway, and propagate the first error found. Would it be simpler to just use `.try_collect` here?

Reply:
You are right, I managed to simplify it further!